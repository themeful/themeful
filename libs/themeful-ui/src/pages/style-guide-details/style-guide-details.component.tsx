import { propertySelect } from '@properties'
import { Component, Event, EventEmitter, h, Host, Prop, State } from '@stencil/core'
import {
  ExtendedStyle,
  ExtendedStyleGuide,
  ExtendedStyleGuides,
  FormIntegrationActions,
  StyleGroup,
  StyleGuideBase,
  StyleTypeGroup,
} from '@typings'
import { Observable, Subject, Subscription } from 'rxjs'

@Component({
  tag: 'tf-style-guide-details',
  styleUrl: 'style-guide-details.component.scss',
  shadow: true,
})
export class StyleGuideDetailsComponent {
  /** Style Guides */
  @Prop() styleGuides$: Observable<ExtendedStyleGuides>

  /** Style Guide Slug */
  @Prop() match: { params: { slug: string } }

  @State() styleGuide: ExtendedStyleGuide
  private formData$ = new Subject()

  /** Event emitted when an action is triggered */
  @Event({ composed: false }) action: EventEmitter<FormIntegrationActions>

  private onAction = ({ detail }): void => {
    if (detail.action !== 'close') {
      this.action.emit(detail)
    }
  }

  private sub = new Subscription()
  private styleGuideGroups: { [styleguide: string]: string[] }

  private openStyleGuideForm = (styleGuide?: string, styleGuideBase?: StyleGuideBase): void => {
    this.formData$.next({
      form: 'styleguide',
      identifier: styleGuide,
      fields: styleGuideBase,
    })
  }

  private openStyleForm = (styleGuide: string, extendedStyle?: ExtendedStyle): void => {
    this.formData$.next({
      form: 'style',
      identifier: { styleGuide, style: extendedStyle?.slug },
      groups: this.styleGuideGroups[styleGuide],
      propertyTypes: propertySelect,
      fields: extendedStyle,
    })
  }

  public componentWillLoad(): void {
    this.sub.add(
      this.styleGuides$?.subscribe((styleGuides) => {
        styleGuides.forEach((styleguide) => {
          console.log('3', styleguide, this.match.params.slug)
          if (styleguide.slug === this.match.params.slug) {
            this.styleGuide = styleguide
            console.log('2', this.styleGuide)
          }
        })
        console.log('1', this.styleGuide)
        this.styleGuideGroups = styleGuides.reduce((result, styleguide) => {
          result[styleguide.slug] = []
          styleguide.types.forEach((type) => {
            type.groups.forEach((group) => {
              result[styleguide.slug].push(group.name)
            })
          })
          return result
        }, {})
      })
    )
  }

  public disconnectedCallback(): void {
    this.sub.unsubscribe()
  }

  public render(): HTMLTfStyleGuidesElement {
    if (!this.styleGuide) {
      return
    }
    return (
      <Host>
        <stencil-route-title pageTitle={`Themeful - ${this.styleGuide.name}`} />
        <tf-navigation
          items={[
            {
              label: 'Edit Style Guide',
              callback: () =>
                this.openStyleGuideForm(this.styleGuide.slug, {
                  name: this.styleGuide.name,
                  baseFontSize: this.styleGuide.baseFontSize,
                }),
            },
            {
              label: 'Add Style',
              callback: () => this.openStyleForm(this.styleGuide.slug),
            },
          ]}
          size="small"
        />
        <main class="style-guide__wrapper">{this.renderStyleGuide(this.styleGuide)}</main>
        <tf-form-integration {...{ formData$: this.formData$, onAction: this.onAction }} />
      </Host>
    )
  }

  private renderStyleGuide(styleGuide: ExtendedStyleGuide): HTMLElement {
    return (
      <div class="style-guide">
        <h2>{styleGuide.name}</h2>
        {styleGuide.types && styleGuide.types.map((type) => this.renderType(type, styleGuide.slug))}
      </div>
    )
  }

  private renderType(type: StyleTypeGroup, styleGuideSlug: string): HTMLElement {
    return (
      <div class="style-guide__type">
        <h3>{type.name}</h3>
        <div class="style-guide__groups">
          {type.groups &&
            type.groups.map((group) =>
              this.renderGroup(group, styleGuideSlug, type.groups.length > 1)
            )}
        </div>
      </div>
    )
  }

  private renderGroup(group: StyleGroup, styleGuideSlug: string, showGroup: boolean): HTMLElement {
    return (
      <div class="style-guide__group property--grouped">
        {showGroup && <h5>{group.name}</h5>}
        {group.styles && group.styles.map((style) => this.renderStyle(style, styleGuideSlug))}
      </div>
    )
  }

  private renderStyle(extendedStyle: ExtendedStyle, styleGuide: string): HTMLTfPropertyElement {
    return (
      <tf-property
        {...{
          extendedStyle,
          showGroup: false,
          onEdit: () => this.openStyleForm(styleGuide, extendedStyle),
        }}
      />
    )
  }
}
