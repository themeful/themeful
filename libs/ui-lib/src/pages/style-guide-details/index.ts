export * from './style-guide-details.component'
