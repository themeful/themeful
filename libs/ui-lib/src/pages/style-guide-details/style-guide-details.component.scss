:host {
  display: block;
}

.style-guide {
  &__wrapper {
    padding: 0 16px;
  }
  &__type {
    h3 {
      margin: 0;
      padding: 8px 8px 8px 0;
    }
  }
  &__groups {
    display: flex;
    gap: 16px;
  }
  &__group {
    text-align: center;
    border: 1px solid currentColor;
    border-radius: 8px;
    overflow: hidden;
    margin-bottom: 16px;

    h5 {
      font: var(--tf-default-font);
      border-bottom: 1px solid currentColor;
      margin: 0;
      padding: 8px;
    }

    tf-property:not(:last-of-type) {
      box-shadow: 0 1px 0 0 rgba(200, 200, 200, 0.55);
    }
  }
}
