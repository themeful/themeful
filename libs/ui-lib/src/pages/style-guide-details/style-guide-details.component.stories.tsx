import { h } from '@stencil/core'
import { action } from '@storybook/addon-actions'
import { of } from 'rxjs'
import sample from './style-guide-details.sample.json'

export default {
  title: 'Pages/Style Guide Details',
  args: {
    styleGuides$: of(sample),
    match: { params: { slug: 'styleGuide1' } },
    onAction: action('triggered action'),
  },
}

export const styleGuideDetails = (args) => {
  return (
    <div>
      <div class="header">
        <h1>Style Guide Details</h1>
      </div>
      <div class="content">
        <div style={{ position: 'relative' }}>
          <tf-style-guide-details {...args} />
        </div>
      </div>
    </div>
  )
}
