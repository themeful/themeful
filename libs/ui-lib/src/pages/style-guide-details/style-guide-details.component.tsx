import { propertySelect } from '@properties'
import { Component, Event, EventEmitter, h, Host, Prop, State } from '@stencil/core'
import { RouterHistory } from '@stencil/router'
import {
  ExtendedStyle,
  ExtendedStyleGuide,
  ExtendedStyleGuides,
  FormIntegrationActions,
  StyleGroup,
  StyleGuideBase,
  StyleGuideIntegrationAction,
  StyleTypeGroup,
} from '@typings'
import { Observable, Subject, Subscription } from 'rxjs'
import { Components } from '../../components'
import '../../components/button'
import '../../components/navigation'
import '../../components/property'
import '../../forms/form-integration'

@Component({
  tag: 'tf-style-guide-details',
  styleUrl: 'style-guide-details.component.scss',
  shadow: true,
})
export class StyleGuideDetailsComponent {
  /** History */
  @Prop() history!: RouterHistory

  /** Style Guides */
  @Prop() styleGuides$!: Observable<ExtendedStyleGuides>

  /** Style Guide Slug */
  @Prop() match!: { params: { slug: string } }

  @State() styleGuide!: ExtendedStyleGuide
  private formData$ = new Subject()

  /** Event emitted when an action is triggered */
  @Event({ composed: false }) action!: EventEmitter<FormIntegrationActions>

  private onAction = ({ detail }: { detail: StyleGuideIntegrationAction }): void => {
    if (detail.action !== 'close') {
      this.action.emit(detail)
      if (['duplicate', 'delete'].includes(detail.action)) {
        this.history.push(`/styleguides`, {})
      }
    }
  }

  private sub = new Subscription()
  private styleGuideGroups!: { [styleguide: string]: string[] }

  private openStyleGuideForm = (styleGuide?: string, styleGuideBase?: StyleGuideBase): void => {
    this.formData$.next({
      form: 'styleguide',
      identifier: styleGuide,
      fields: styleGuideBase,
    })
  }

  private openStyleGuideDuplicateForm = (styleGuide: string, name: string): void => {
    this.formData$.next({
      form: 'styleguideDuplicate',
      identifier: styleGuide,
      fields: { name },
    })
  }

  private openStyleForm = (styleGuide: string, extendedStyle?: ExtendedStyle): void => {
    this.formData$.next({
      form: 'style',
      identifier: { styleGuide, style: extendedStyle?.slug },
      groups: this.styleGuideGroups[styleGuide],
      propertyTypes: propertySelect,
      fields: extendedStyle,
    })
  }

  public componentWillLoad(): void {
    this.sub.add(
      this.styleGuides$?.subscribe((styleGuides) => {
        styleGuides.forEach((styleguide) => {
          if (styleguide.slug === this.match.params.slug) {
            this.styleGuide = styleguide
          }
        })
        this.styleGuideGroups = styleGuides.reduce(
          (result: { [key: string]: string[] }, styleguide) => {
            result[styleguide.slug] = []
            styleguide.types.forEach((type) => {
              type.groups.forEach((group) => {
                result[styleguide.slug].push(group.name)
              })
            })
            return result
          },
          {}
        )
      })
    )
  }

  public disconnectedCallback(): void {
    this.sub.unsubscribe()
  }

  public render(): HTMLTfStyleGuidesElement {
    return (
      this.styleGuide && (
        <Host>
          <stencil-route-title pageTitle={`Themeful - ${this.styleGuide.name}`} />
          <tf-navigation
            items={[
              {
                label: 'Edit Style Guide',
                callback: () =>
                  this.openStyleGuideForm(this.styleGuide.slug, {
                    name: this.styleGuide.name,
                    baseFontSize: this.styleGuide.baseFontSize,
                  }),
              },
              {
                label: 'Add Style',
                callback: () => this.openStyleForm(this.styleGuide.slug),
              },
              {
                label: 'Duplicate Style Guide',
                callback: () =>
                  this.openStyleGuideDuplicateForm(this.styleGuide.slug, this.styleGuide.name),
              },
            ]}
            size="small"
          />
          <div class="style-guide__wrapper">{this.renderStyleGuide(this.styleGuide)}</div>
          <tf-form-integration
            {...({
              formData$: this.formData$,
              onAction: this.onAction,
            } as Components.TfFormIntegration)}
          />
        </Host>
      )
    )
  }

  private renderStyleGuide(styleGuide: ExtendedStyleGuide): HTMLElement {
    return (
      <div class="style-guide">
        <h2>{styleGuide.name}</h2>
        {styleGuide.types && styleGuide.types.map((type) => this.renderType(type, styleGuide.slug))}
      </div>
    )
  }

  private renderType(type: StyleTypeGroup, styleGuideSlug: string): HTMLElement {
    return (
      <div class="style-guide__type">
        <h3>{type.name}</h3>
        <div class="style-guide__groups">
          {type.groups &&
            type.groups.map((group) =>
              this.renderGroup(group, styleGuideSlug, type.groups.length > 1)
            )}
        </div>
      </div>
    )
  }

  private renderGroup(group: StyleGroup, styleGuideSlug: string, showGroup: boolean): HTMLElement {
    return (
      <div class="style-guide__group property--grouped">
        {showGroup && <h5>{group.name}</h5>}
        {group.styles && group.styles.map((style) => this.renderStyle(style, styleGuideSlug))}
      </div>
    )
  }

  private renderStyle(extendedStyle: ExtendedStyle, styleGuide: string): HTMLTfPropertyElement {
    return (
      <tf-property
        {...{
          extendedStyle,
          showGroup: false,
          onEdit: () => this.openStyleForm(styleGuide, extendedStyle),
        }}
      />
    )
  }
}
