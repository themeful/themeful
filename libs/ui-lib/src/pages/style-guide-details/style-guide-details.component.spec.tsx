import { newSpecPage } from '@stencil/core/testing'
import { StyleGuideDetailsComponent } from './style-guide-details.component'

describe('StyleGuideDetailsComponent', () => {
  it('renders', async () => {
    const { root } = await newSpecPage({
      components: [StyleGuideDetailsComponent],
      html: '<tf-style-guide-details></tf-style-guide-details>',
      url: 'http://localhost/',
    })
    expect(root).toMatchSnapshot()
  })

  it('renders with values', async () => {
    const { root } = await newSpecPage({
      components: [StyleGuideDetailsComponent],
      html: `<tf-style-guide-details></tf-style-guide-details>`,
      url: 'http://localhost/',
    })
    expect(root).toMatchSnapshot()
  })
})
