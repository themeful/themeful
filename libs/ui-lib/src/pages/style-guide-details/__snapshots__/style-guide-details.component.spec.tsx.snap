// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StyleGuideDetailsComponent renders 1`] = `
<tf-style-guide-details>
  <mock:shadow-root></mock:shadow-root>
</tf-style-guide-details>
`;

exports[`StyleGuideDetailsComponent renders with values 1`] = `
<tf-style-guide-details>
  <mock:shadow-root></mock:shadow-root>
</tf-style-guide-details>
`;
